### DISCUSSION TITLE
System.InvalidCastException: 'Object must implement IConvertible.' when invoking callback function with object return value

### DISCUSSION THREAD

**Discussion by jsiemensLatitudegeo** [sebastienros/jint#1534] [Q&A]
**Version used**

Tried both 2.11.58 and 3.0.0-beta-204.

**Describe the bug**

I'm trying to implement a function that returns a value, and when the value is an object, I get an InvalidCastException.

**To Reproduce**

        [Fact]
        public void TestJint2()
        {
            var flatMap = (IEnumerable<object> list, Func<object, int, IEnumerable<object>, object> callback) =>
            {
                int i = 0;
                return list.SelectMany(x =>
                {
                    var intermediate = callback(x, i++, list);
                    if (intermediate is IEnumerable enumerable)
                    {
                        return enumerable.OfType<object>();
                    }

                    return new[] { intermediate };
                }).ToArray();
            };

            var engine = new Jint.Engine()
                .SetValue("flatMap", flatMap)
                .Execute(
                        @"var items = flatMap([1, [2, 3]], function(x) { 
                            return x; 
                        });");
        }

**Expected behavior**

The callback function is called, and the object is returned.

**Additional context**

When the callback is called with a primitive value like an integer, and an integer is returned, it works fine. It fails for reference types like arrays.

I was trying to implement the JavaScript flatMap method:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap

**Comment by lahma** [comment]
I think that your `flatMap` signature is just too complex for Jint to handle. There are a lot samples which kind of signatures will work in InteropTests*.cs in the repo, maybe you could adapt something from there.

Generally if you want to implement JavaScript functionality, the signature should be `JsValue FlatMap(JsValue thisObj, JsValue[] arguments)`, like we have for the [already implemented Array.flatMap](https://github.com/sebastienros/jint/blob/8b0da8ad3a754b4bb5d7da9d2492acef7ab25aa4/Jint/Native/Array/ArrayPrototype.cs#L546-L565).
